### PR TITLE
Feature/event notification

### PR DESCRIPTION
**Pull request by jacquesporveau**
- Users should be able to configure how many days before a presentation the presenter is notified of their talk.
 - The configuration should apply to all upcoming presentations

### CODE REVIEW

File: app/models/event.rb
Comment (qr8r) [inline app/models/event.rb]: Consider making this an AR scope instead of a class method.

File: spec/models/event_spec.rb
Comment (qr8r) [inline spec/models/event_spec.rb]: Typo?

File: spec/models/event_spec.rb
Comment (qr8r) [inline spec/models/event_spec.rb]: Your spec should have a `subject` defined. the subject helper takes an optional symbol param allowing you to make more semantic expectations 
```ruby
subject(:smarties) { ... }

it 'sdfsdfsdf' do
  expect(:smarties).to be_better_than :m_and_ms
end
```

File: spec/models/event_spec.rb
Comment (qr8r) [inline spec/models/event_spec.rb]: Rspec allows you to dynamically use predicate methods with the `be` helper. 

`expect([]).to be_empty`

File: app/models/event.rb
Comment (qr8r) [inline app/models/event.rb:1]: This commit needs a commit message explaining why this change is required.

File: app/helpers/email_helper.rb
Comment (qr8r) [inline app/helpers/email_helper.rb]: This class strikes me as a Job. 

Consider inheriting from ApplicationJob

File: app/helpers/email_helper.rb
Comment (qr8r) [inline app/helpers/email_helper.rb]: This method might not be needed if you inherit from ApplicationJob

File: app/jobs/event_notification_job.rb
Comment (qr8r) [inline app/jobs/event_notification_job.rb]: Unneeded helper statement. The EmailHelper has no public api, so declaring it as a helper here does nothing

File: app/jobs/event_notification_job.rb
Comment (qr8r) [inline app/jobs/event_notification_job.rb]: I think this class should be doing what your Email helper is doing.

**Review by qr8r — CHANGES_REQUESTED** [review]

File: app/models/event.rb
Comment (jacquesporveau) [reply app/models/event.rb]: Hey I didn't know about that syntax. Very cool. Thanks for the tip.

File: spec/models/event_spec.rb
Comment (qr8r) [inline spec/models/event_spec.rb]: No need to change this, is good to know that when you are referencing the class that you are testing in RSpec, you can use the `described_class` helper.

Makes it easier to change the name of the class if you ever had to

**Review by qr8r — COMMENTED** [review]
Looks good. Will approve once #19 is merged

File: spec/models/event_spec.rb
Comment (jacquesporveau) [reply spec/models/event_spec.rb]: Will keep that in mind.

File: spec/models/event_spec.rb
Comment (gregdaynes) [inline spec/models/event_spec.rb:75]: Double negative - reads funny.
it 'only includes events that require notification' or something? minor, and doesn't really matter.

File: app/models/event.rb
Comment (gregdaynes) [inline app/models/event.rb:3]: Unfamiliar with Rails, but this would be beneficial to be defined in a constants/config file for easier future changes.

`config/application.rb`
then for dev/testing purposes, overload it in `confing/environemnts/`

Comment (gregdaynes) [inline app/models/event.rb]: Nevermind. I guess that's against common Rails practive.

Comment (jacquesporveau) [reply app/models/event.rb:3]: Totally. We are adding the Figaro gem in the oauth work that I am doing. Once that is configured I think that's a great idea.

File: app/jobs/event_notification_job.rb
Comment (graygilmore) [inline app/jobs/event_notification_job.rb]: I think we should remove the commented out code. Instead, I'd prefer to add a `#TODO` comment to `presenter` above, detail that this is a stubbed model for now.

File: app/views/person_mailer/presentation_notification.html.erb
Comment (graygilmore) [inline app/views/person_mailer/presentation_notification.html.erb]: I think we can add actual content here since we're just stubbing out the model that we're going to be getting anyways.

File: app/mailers/person_mailer.rb
Comment (graygilmore) [inline app/mailers/person_mailer.rb]: Maybe `subject: "You have a presentation in #{n} days"`?

**Review by graygilmore — CHANGES_REQUESTED** [review]

File: app/jobs/event_notification_job.rb
Comment (jacquesporveau) [reply app/jobs/event_notification_job.rb]: With @gregdaynes stuff being merged I think this is ready to code now so instead of leaving this stubbed I'll just finish it.

Comment (jacquesporveau) [reply app/jobs/event_notification_job.rb]: This has been completed in 1b60260

File: app/jobs/event_notification_job.rb
Comment (graygilmore) [inline app/jobs/event_notification_job.rb]: When does this run?

**Review by graygilmore — CHANGES_REQUESTED** [review]
I would like to see the last commit squashed into the one before it.

File: app/jobs/event_notification_job.rb
Comment (jacquesporveau) [reply app/jobs/event_notification_job.rb]: Right now it doesn't we haven't added in Redis or another task runner yet.

Comment (qr8r) [reply app/jobs/event_notification_job.rb]: I will add stories to implement a scheduling provider and a background processing queue provider.
We wont actually be running this stuff until we set up a staging environment

For now this only requires a rake task to manually run the job from the console. In the Development environment by default Active job will run jobs inline, until we implement a background processing queue

Comment (jacquesporveau) [reply app/jobs/event_notification_job.rb]: rake task has been introduced in dfb704af07ab8701537e45a13d65bef87cdf8dd0

**Comment by qr8r** [comment]
What does `RFR` mean in the description? 
Ready for review.

File: app/jobs/event_notification_job.rb
Comment (qr8r) [inline app/jobs/event_notification_job.rb]: Unnecessary variable assignment. `event`  is still available in the nested loop and typing `event_date` doesn't save much space.

File: app/jobs/event_notification_job.rb
Comment (qr8r) [inline app/jobs/event_notification_job.rb]: Probably want to eager load presentations and person here so that they are all fetch in the same query. This way we avoid creating an N+1 query to the database. 

If there were a large number of events involved, we might want to consider using the `find_each` AR method rather than a simple each. but for now its probably fine

File: spec/factories.rb
Comment (qr8r) [inline spec/factories.rb]: What is this being changed? There is a unique validation on emails making sequences a little clunky if you were ever going to use the factories in the dev environment

File: spec/jobs/event_notification_job_spec.rb
Comment (qr8r) [inline spec/jobs/event_notification_job_spec.rb]: described class.perform should probably be your subject since thats what you are "expecteing" in both tests

File: app/jobs/event_notification_job.rb
Comment (qr8r) [inline app/jobs/event_notification_job.rb:1]: This commit is too large. The mailer and the job should be in separate commits

File: lib/tasks/send_presentation_notifications.rake
Comment (qr8r) [inline lib/tasks/send_presentation_notifications.rake]: So it took me until now to realize this, but our job doesnt take any arguments. Which means we can't take advantage of splitting our event notifications out into multiple processes to make them faster.

I propose making EventNotificationJob#perform accept an event as an argument. 
It will be the rake tasks responsibility to find all events that need to be notified, and then initialize a new job for every event.

That way when we implement our background processing. The queue can run multiple jobs at the same time, instead of having to notify every event from within the same job

**Review by qr8r — CHANGES_REQUESTED** [review]
Getting there!

File: spec/factories.rb
Comment (jacquesporveau) [reply spec/factories.rb]: This doesn't work in the test environment. If you attempt to use this style of get a Validation failed: Email has already been taken error. I've googled a bunch and tried to debug it and every source I find recommends doing things this way so I tried it and it solved the test environment issues I was having.

File: app/jobs/event_notification_job.rb
Comment (jacquesporveau) [reply app/jobs/event_notification_job.rb:1]: Got it!

**Review by graygilmore — APPROVED** [review]

File: spec/factories.rb
Comment (jacquesporveau) [reply spec/factories.rb]: Alright found the solution for this.

https://github.com/thoughtbot/factory_girl/blob/master/GETTING_STARTED.md#dynamic-attributes

**Comment by jacquesporveau** [comment]
Weird. Had to reset branch set to merge into to re-check for merge conflicts as it was picking up some remotely but not locally.

File: app/models/event.rb
Comment (qr8r) [inline app/models/event.rb]: The `[]` around person here is not needed.

In this case I wouldn't preload the associations here (unless the scope itself needed it for some reason). you take this scope from being 1 query to 3, and you might not always want the associated records.

Better to leave the `includes` call to the person who will actually be using the scope

File: spec/factories.rb
Comment (qr8r) [inline spec/factories.rb:6]: :+1:

File: lib/tasks/send_presentation_notifications.rake
Comment (qr8r) [inline lib/tasks/send_presentation_notifications.rake]: You probably want perform_later here. When there is no processing background in place, rails will always perform jobs inline, even if you asked it to `perform_later`.

This way we dont need to remember to change this when we add a processing background

**Review by qr8r — CHANGES_REQUESTED** [review]
Almost!!!

File: app/models/event.rb
Comment (jacquesporveau) [reply app/models/event.rb]: Hmmmm right I see where you are coming from however the instance in which I need to do that preloading is inside a rake task. I am not sure if we want that much logic in there?

```
require 'rake'

task send_presentation_notifications: :environment do
  Event.requiring_notification.each do
    EventNotificationJob.perform(event)
  end
end
```

That's what it's looking like right now. I could put the eager load in there but I just remember you mentioning to stay away from logic inside rake tasks.

File: spec/factories.rb
Comment (jacquesporveau) [reply spec/factories.rb:6]: YUS

File: lib/tasks/send_presentation_notifications.rake
Comment (jacquesporveau) [reply lib/tasks/send_presentation_notifications.rake]: Is that some Rails magic or something? I have a method defined as perform so I assumed that's what I had to call. Does it know about perform_later? 

Just looked up the docs. Yes it does? Weird.

File: app/models/event.rb
Comment (jacquesporveau) [reply app/models/event.rb]: Alright got this one figured out. The scope in the event model is just for itself and in the event job I have...
```
class EventNotificationJob < ApplicationJob
  def self.perform(event)
    event.presentations.joins(:person).each do |presentation|
      PersonMailer.presentation_notification(
        presentation.person,
        presentation.event.date
      ).deliver
    end
  end
end
```

Comment (jacquesporveau) [reply app/models/event.rb]: Tried this in console and it grabs only presentations that have a person as well as loading the people in memory.

File: lib/tasks/send_presentation_notifications.rake
Comment (qr8r) [inline lib/tasks/send_presentation_notifications.rake]: Have you been able to run this rake task from the console?

File: spec/factories.rb
Comment (qr8r) [inline spec/factories.rb:18]: :+1: Came in handy when I wanted to test this locally. Good call

**Review by qr8r — CHANGES_REQUESTED** [review]

File: lib/tasks/send_presentation_notifications.rake
Comment (jacquesporveau) [reply lib/tasks/send_presentation_notifications.rake]: Sry man that's my bad. I should have tested this more in depth. Made the missing var change.

**Review by qr8r — APPROVED** [review]

**Pull request merged**
